crates/network/src/diff_serialization.rs

```diff
@@ -54,3 +75,675 @@ impl WorldDiffDeduplicator {
+        }
+        Ok(entity)
+    }
+
+    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::MapAccess<'de>,
+    {
+        let mut entity = Entity::new();
+        while let Some(desc) = map.next_key_seed(NetworkedComponentDescVisitor {
+            known_component_paths: self.known_component_paths,
+        })? {
+            let Some(ser) = desc.attribute::<Serializable>() else {
+                return Err(serde::de::Error::custom(format!("tried to deserialize non-serializable component {:?}", desc)));
+            };
+            let entry = map.next_value_seed(ser.deserializer(desc))?;
+            entity.set_entry(entry);
+        }
+        Ok(entity)
+    }
+}
+
+impl<'a, 'de> serde::de::DeserializeSeed<'de> for NetworkedEntityVisitor<'a> {
+    type Value = Entity;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(self)
+    }
+}
+
+struct NetworkedComponentDescsVisitor<'a> {
+    known_component_paths: &'a HashMap<u32, String>,
+}
+
+impl<'a, 'de> serde::de::Visitor<'de> for NetworkedComponentDescsVisitor<'a> {
+    type Value = Vec<ComponentDesc>;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("sequence of networked ComponentDesc")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        let mut components = Vec::with_capacity(seq.size_hint().unwrap_or_default());
+        while let Some(desc) = seq.next_element_seed(NetworkedComponentDescVisitor {
+            known_component_paths: self.known_component_paths,
+        })? {
+            components.push(desc);
+        }
+        Ok(components)
+    }
+}
+
+impl<'a, 'de> serde::de::DeserializeSeed<'de> for NetworkedComponentDescsVisitor<'a> {
+    type Value = Vec<ComponentDesc>;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(self)
+    }
+}
+
+struct NetworkedComponentDescVisitor<'a> {
+    known_component_paths: &'a HashMap<u32, String>,
+}
+
+impl<'a, 'de> serde::de::Visitor<'de> for NetworkedComponentDescVisitor<'a> {
+    type Value = ComponentDesc;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("u32 component index")
+    }
+
+    fn visit_u32<E>(self, v: u32) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        match self.known_component_paths.get(&v) {
+            Some(path) => {
+                let component = with_component_registry(|r| r.get_by_path(path));
+                match component {
+                    Some(desc) => Ok(desc),
+                    None => Err(serde::de::Error::custom(format!("No such component: {v}"))),
```

Comment:
Is there any sane path in which this can be triggered.

My only idea is if receiving and deserialization occurs before the *ServerInfo* recv future has been completed

Reply:
I mostly followed the pattern of the regular serialization implementation.

I'd say it can either be caused by broken deserialization (for example subtle error when mismatched versions are used) or some timing issue as you mentioned.